packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -797,14 +771,44 @@ const VirtualListBaseFactory = (type) => {
 			delete rest.spacing;
 			delete rest.updateStatesAndBounds;
 
+			const
+				{dataSize, direction, itemSize, overhang, rtl, spacing} = this.prevProps,
+				{props} = this,
+				hasMetricsChanged = (
+					direction !== props.direction ||
+					((itemSize instanceof Object) ? (itemSize.minWidth !== props.itemSize.minWidth || itemSize.minHeight !== props.itemSize.minHeight) : itemSize !== props.itemSize) ||
+					overhang !== props.overhang ||
+					spacing !== props.spacing
+				);
+
+			this.hasDataSizeChanged = (dataSize !== props.dataSize);
+
+			if (hasMetricsChanged || this.props.clientSize && this.primary === null && this.secondary === null) {
```

Comment:
All these conditions seem to happen after the first render. IOW, metrics and data size change would only happen after its first mount and subsequent renders.

I wonder if there are any noticeable perf loss if these changes were pushed back to `componentDidUpdate` phase. By doing so, I think we can remove the hacky `updatePrevProps()` and (yet another) the internal member variable. Plus, we can remove the direct `state` mutation hack that I pointed out above.

Reply:
The following condition is for the first render.
- this.props.clientSize && this.primary === null && this.secondary === null

The original code  for the condition above was in `UNSAFE_componentWillMount ()`. The reason why we added the condition was to reduce calling `render()` from twice to once to enhance app launch time. The impact for the patch was so big. Please check the comments in PLAT-37236. If it is ok to call `render()` more, we could just undo the patch by removing the condition. But I'm afraid of increasing the launch time.

Reply:
Generic questions:

- Have we promoted `clientSize` props over setting dimensions to developers? I don't see any apps using this, unfortunately :(
- I don't think we can't ignore what happens when VirtualList is used with `Scroller` as `VirtualListBase` has a strong relationship with `Scrollable`. And `syncClientSize` gets called when `Scrollable` updates.
- If the goal is to minimize the initial render when `clientSize` were provided as a prop, can we calculate metrics in `constructor` instead? Essentially we're initializing the metrics via `calculateMetrics` so that we can position items on render.

Reply:
> * Have we promoted `clientSize` props over setting dimensions to developers? I don't see any apps using this, unfortunately :(

I found just only one use case in Settings app. Actually the issue related with `clientSize` was raised from Settings Picture panel. But the `clientSize` in the panel was not used.
\com.webos.app.settings\src\views\General\LanguageList.js

```
	setVirtualListProps (listData, comp) {
		const retProps = {
			itemSize: ri.scale(78),
			dataSize: listData.length,
			isItemDisabled: (index) => (listData[index].disabled),
			itemRenderer: comp
		};

		if (typeof window === 'object' && window.testing) {
			retProps.clientSize = {
				clientWidth: 1920,
				clientHeight: 1080
			};
		}

		return retProps;
	}

	const virtualListProps = this.setVirtualListProps(listData, this.getLanguageListItem());

	<VirtualList {...virtualListProps} className={css['general-language-list-virtuallist']} />
```

Reply:
> * If the goal is to minimize the initial render when `clientSize` were provided as a prop, can we calculate metrics in `constructor` instead? Essentially we're initializing the metrics via `calculateMetrics` so that we can position items on render.

Yes. I think it makes sense. As you proposed the code in [your branch](https://github.com/enactjs/enact/tree/feature/PLAT-72723-stephen) I'll try to apply it.

Reply:
<On Feb 18th>
- I move the code in `UNSAFE_componentWillMount()` to `constructor` related with `clientSize`.
But `updateStatesAndBounds` was renamed as `getStatesAndUpdateBounds` to get a new state because we could not call `setState` in `constructor`.